src/Database/LSMTree/Internal.hs

```diff
@@ -492,6 +530,100 @@ toNormalLookupResult = \case
 
+-- | See 'Database.LSMTree.Normal.updates'.
+updates ::
+     forall m h k v blob.
+     V.Vector (k, Entry v blob)
+  -> TableHandle m h
+  -> m ()
+updates = error "updates: not yet implemented" -- TODO: implement
+
+{-------------------------------------------------------------------------------
+  Snapshots
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE snapshot :: SnapshotName -> TableHandle IO h -> IO Int #-}
+-- |  See 'Database.LSMTree.Normal.snapshot''.
+snapshot ::
+     m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
+  => SnapshotName
+  -> TableHandle m h
+  -> m Int
+snapshot snap th = do
+    withOpenTable th $ \thEnv -> do
+      wb <- readMVar (tableWriteBuffer thEnv)
+      levels <- readMVar (tableLevels thEnv)
+      n <- incrUniqCounter (tablesSessionUniqCounter thEnv)
+      wbRun <- Run.fromWriteBuffer (tableHasFS thEnv) (RunFsPaths (Paths.activeDir (tableSessionRoot thEnv)) n) wb
+      Run.removeReference (tableHasFS thEnv) wbRun
+
+      let levels' = case V.uncons levels of
+            Nothing             -> V.singleton (Level $ V.singleton wbRun)
+            Just (Level rs, tl) -> Level (V.cons wbRun rs) `V.cons` tl
+
+      let runNumbers = V.map (V.map (runNumber . Run.runRunFsPaths) . residentRuns) levels'
+
+      FS.withFile
+        (tableHasFS thEnv)
+        (Paths.snapshot (tableSessionRoot thEnv) snap)
+        (FS.WriteMode FS.MustBeNew) $ \h ->
+          void $ FS.hPutAllStrict (tableHasFS thEnv) h (BSC.pack $ show (runNumbers, tableConfig thEnv))
+
+      pure $! V.sum (V.map V.length runNumbers)
+
+{-# SPECIALISE open :: Session IO h -> SnapshotName -> IO (TableHandle IO h) #-}
+-- |  See 'Database.LSMTree.Normal.open'.
+open ::
+     m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
+  => Session m h
+  -> SnapshotName
+  -> m (TableHandle m h)
+open sesh snap = do
+    (runPaths, conf) <- withOpenSession sesh $ \seshEnv -> do
+        bs <- FS.withFile
+          (sessionHasFS seshEnv)
+          (Paths.snapshot (sessionRoot seshEnv) snap)
+          FS.ReadMode $ \h ->
+            FS.hGetAll (sessionHasFS seshEnv) h
+        let (runNumbers, conf) = read . BSC.unpack . BSC.toStrict $ bs
+            runPaths = V.map (V.map (RunFsPaths (Paths.activeDir $ sessionRoot seshEnv))) runNumbers
+        pure (runPaths, conf)
+    newWithLevels sesh conf $ \hfs -> openLevels hfs runPaths
+
+{-# SPECIALISE openRun :: HasFS IO h -> RunFsPaths -> Cont (IO a) (Run (FS.Handle h)) #-}
+-- | Open a run.
+--
+-- If an error occurs when opening multiple runs in sequence, then we have to
+-- make sure that all runs that have been succesfully opened already are closed
+-- again. The 'Cont'inuation allows us to do so by folding 'bracketOnError' over
+-- the action @m a@.
+openRun ::
+     m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
+  => HasFS m h -> RunFsPaths -> Cont (m a) (Run (FS.Handle h))
```

Comment:
As discussed, there might be a simpler solution. Even just using some function could make it easier to understand what's happening without knowing `Cont` well, also in terms of strictness of `>>=`, which is relevant for `mapMStrict`.

Reply:
I'm pretty happy with the current implementation. I've gone through a few implementations, and this is the most concise and doesn't require us to fiddle with the vectors -- we get the structure for free using the `Cont` monad. I've updated the documentation a bit, and I've added an example in there too

RE the strictness concerns, you're right that the strictness depends on the implementation of bind. To verify whether the resulting levels don't contain thunks, I've added a `NoThunks` assertion that is checked before opening a table:

```haskell
assertNoThunks :: NoThunks a => a -> b -> b
assertNoThunks x = assert p
  where p = case unsafeNoThunks x of
              Nothing -> True
              Just thunkInfo -> error $ "Assertion failed: found thunk" <> show thunkInfo
```
It's slightly hacky because I'm hijacking the `assert` functionality to show a custom `assert` error message. The `Bool` I'm passing to the `assert` contains an `error` itself. :stuck_out_tongue: The nice thing is that the `error` is erased when compiled with optimisations

Reply:
Example: when I remove the `$!` from `mapMStrict`
```diff
-mapMStrict f v = V.mapM (f >=> (pure $!)) v
+mapMStrict f v = V.mapM f v
```
Then this test shows a `NoThunks` assertion error
```
>  cabal run lsm-tree-test -- -p "interim"
lsm-tree
  Test.Database.LSMTree.Internal
    prop_interimOpenTable: OK (0.01s)
      +++ OK, failed as expected. (after 1 test and 8 shrinks):
      Exception:
        Assertion failed: found thunkThunkInfo {thunkInfo = Left ["Boxed.Vector","Boxed.Vector"]}
        CallStack (from HasCallStack):
          error, called at src/Database/LSMTree/Internal/Assertions.hs:43:33 in lsm-tree-0.1.0.0-inplace:Database.LSMTree.Internal.Assertions
      InMemLookupData {runData = fromList [(SerialisedKey [0,0,0,0,0,0],Delete)], lookups = []}
```

Reply:
Nice trick with the `assert`!

Question on exception handling: Do we need to include constructing the table handle (once we have the levels of loaded runs) into the `bracket` or can we do that outside? Similarly, if removing one reference (potentially closing a run) throws, should we continue trying to close the others or can we just give up?

I have one alternative implementation I have in mind that especially works well when answering "no" above, but can be adjusted. I think the complications here stem from having two stacks we want to keep simultaneously: the one tracking where we are in constructing the vectors, and the one tracking which runs need to be cleaned up. We could resort to managing the second stack ourselves with something like this:

```haskell
openLevels ::
     m ~ IO
  => HasFS m h
  -> V.Vector (V.Vector RunFsPaths)
  -> m (V.Vector (Level (Handle h)))
openLevels hfs levels = do
    openedRuns <- newIORef []
    -- if something fails, clean up all runs opened so far
    flip onException (traverse_ (Run.removeReference hfs) =<< readIORef openedRuns) $
      flip V.mapMStrict levels $ \l ->
        fmap Level $ flip V.mapMStrict l $ \p -> do
          r <- Run.openFromDisk hfs p
          modifyIORef' openedRuns (r :)
          return r
```

Maybe a bit more straightforward to see what's going on?

Reply:
> Question on exception handling: ...

I think "yes" to both questions. Before proper snapshots are implemented, so in our current situation, runs are created, but never hard-linked or deleted. So, we'd run the risk of leaking file handles if we don't close runs properly, and with proper snapshots we might link leak (hard-links to) files. Hopefully GHC cleans the file descriptors up if we don't close the runs manually, but there is no guarantee even if the file handles have a finaliser attached. And when we start caring about linking/deleting files too, then those are going to leak regardless. The best thing is to just be extra sure that we clean up resources